### ISSUE TITLE
can't build gem on aarch64/Android/Termux/Ubuntu srv jammy 22.04LTS w/ Ruby 3.0.2

### ISSUE THREAD

**Issue by libweirdness** [sparklemotion/sqlite3-ruby#322]
this is my learning-programming environment:
aarch / Android / Termux / Ubuntu server jammy aka 22.04LTS
Ruby 3.0.2

I tried the git version of rbenv,
and discovered that NO version of Ruby before the 3.1.x branch can be built on this machine,
but that one could.
( so how exactly did Ubuntu build 3.0.2?
did they cross-compile it, or is there some bug in the within-termux version? )
bugs.ruby-lang closed my bug report almost immediately, saying it was an "ssl bug".

Now you know the context.

This looks like some C problems, though.

gcc is 11.2, iirc, and clang-13 also is installed.

stderr gives:
```
  1 ERROR:  Error installing sqlite3:                                                               
  2    ERROR: Failed to build gem native extension.                                                 
  3                                                                                                 
  4     current directory: /var/lib/gems/3.0.0/gems/sqlite3-1.4.2/ext/sqlite3                       
  5 /usr/bin/ruby3.0 -I /usr/lib/ruby/vendor_ruby -r ./siteconf20220609-3250-7c0utr.rb extconf.rb   
  6 checking for sqlite3.h... yes
  7 checking for pthread_create() in -lpthread... yes
  8 checking for -ldl... yes
  9 checking for sqlite3_libversion_number() in -lsqlite3... yes                                   
 10 checking for rb_proc_arity()... yes                                                            
 11 checking for rb_integer_pack()... yes                                                          
 12 checking for sqlite3_initialize()... yes                                                       
 13 checking for sqlite3_backup_init()... yes                                                      
 14 checking for sqlite3_column_database_name()... yes                                             
 15 checking for sqlite3_enable_load_extension()... yes                                            
 16 checking for sqlite3_load_extension()... yes                                                   
 17 checking for sqlite3_open_v2()... yes                                                          
 18 checking for sqlite3_prepare_v2()... yes                                                       
 19 checking for sqlite3_int64 in sqlite3.h... yes                                                 
 20 checking for sqlite3_uint64 in sqlite3.h... yes                                                
 21 creating Makefile                                                                              
 22                                                                                                
 23 current directory: /var/lib/gems/3.0.0/gems/sqlite3-1.4.2/ext/sqlite3                          
 24 make DESTDIR\= clean                                                                           
 25                                                                                                
 26 current directory: /var/lib/gems/3.0.0/gems/sqlite3-1.4.2/ext/sqlite3                          
 27 make DESTDIR\=                                                                                 
 28 compiling aggregator.c                                                                         
 29 compiling backup.c                                                                             
 30 compiling database.c                                                                           
 31 database.c: In function ‘exec_batch’:
 32 database.c:726:57: warning: passing argument 3 of ‘sqlite3_exec’ from incompatible pointer type [-Wincompatible-pointer-types]
 33   726 |     status = sqlite3_exec(ctx->db, StringValuePtr(sql), hash_callback_function, callback_ary, &errMsg);                                                                                
 34       |                                                         ^~~~~~~~~~~~~~~~~~~~~~          
 35       |                                                         |                               
 36       |                                                         int (*)(VALUE,  int,  char **,     char **) {aka int (*)(long unsigned int,  int,  char **, char **)}                             
 37 In file included from ./sqlite3_ruby.h:25,                                                     
 38                  from database.c:1:                                                             
 39 /usr/include/sqlite3.h:428:9: note: expected ‘int (*)(void *, int,  char **, char **)’ but argument is of type ‘int (*)(VALUE,  int,  char **, char **)’ {aka ‘int (*)(long unsigned int,  int    ,  char **, char **)’}                                                                          
 40   428 |   int (*callback)(void*,int,char**,char**),  /* Callback function */                    
 41       |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                              
 42 database.c:726:81: warning: passing argument 4 of ‘sqlite3_exec’ makes pointer from integer without a cast [-Wint-conversion]
 43   726 |     status = sqlite3_exec(ctx->db, StringValuePtr(sql), hash_callback_function, callback_ary, &errMsg);
 44       |                                                                                 ^~~~~~~    ~~~~~                                                                                           
 45       |                                                                                 |      
 46       |                                                                                 VALUE {    aka long unsigned int}                                                                          47 In file included from ./sqlite3_ruby.h:25,                                                     
 48                  from database.c:1:                                                            
 49 /usr/include/sqlite3.h:429:3: note: expected ‘void *’ but argument is of type ‘VALUE’ {aka ‘long unsigned int’}                                                                               
 50   429 |   void *,                                    /* 1st argument to callback */            
 51       |   ^~~~~~                                                                               
 52 database.c:728:57: warning: passing argument 3 of ‘sqlite3_exec’ from incompatible pointer type     [-Wincompatible-pointer-types]                                                                
 53   728 |     status = sqlite3_exec(ctx->db, StringValuePtr(sql), regular_callback_function, call    back_ary, &errMsg);                                                                            
 54       |                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~      
 55       |                                                         |                              
 56       |                                                         int (*)(VALUE,  int,  char **,     char **) {aka int (*)(long unsigned int,  int,  char **, char **)}                             
 57 In file included from ./sqlite3_ruby.h:25,                                                     
 58                  from database.c:1:                                                            
 59 /usr/include/sqlite3.h:428:9: note: expected ‘int (*)(void *, int,  char **, char **)’ but argument is of type ‘int (*)(VALUE,  int,  char **, char **)’ {aka ‘int (*)(long unsigned int,  int    ,  char **, char **)’}                                                                         
 60   428 |   int (*callback)(void*,int,char**,char**),  /* Callback function */                   
 61       |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                             
 62 database.c:728:84: warning: passing argument 4 of ‘sqlite3_exec’ makes pointer from integer without a cast [-Wint-conversion]                                                                 
 63   728 |     status = sqlite3_exec(ctx->db, StringValuePtr(sql), regular_callback_function, call    back_ary, &errMsg);                                                                            
 64       |                                                                                    ^~~~    ~~~~~~~~                                                                                              
 65       |                                                                                    |   
 66       |                                                                                    VALUE {aka long unsigned int}                                                                        
 67 In file included from ./sqlite3_ruby.h:25,
 68                  from database.c:1:
 69 /usr/include/sqlite3.h:429:3: note: expected ‘void *’ but argument is of type ‘VALUE’ {aka ‘lon    g unsigned int’}
 70   429 |   void *,                                    /* 1st argument to callback */            
 71       |   ^~~~~~                                                                               
 72 compiling exception.c                                                                          
 73 compiling sqlite3.c                                                                            
 74 compiling statement.c                                                                          
 75 linking shared-object sqlite3/sqlite3_native.so                                                
 76                                                                                                
 77 current directory: /var/lib/gems/3.0.0/gems/sqlite3-1.4.2/ext/sqlite3                          
 78 make DESTDIR\= install                                                                         
 79 make: /usr/bin/mkdir: No such file or directory                                                
 80 make: *** [Makefile:202: .sitearchdir.-.sqlite3.time] Error 127                                
 81                                                                                                
 82 make install failed, exit code 2                                                               
 83                                                                                                
 84 Gem files will remain installed in /var/lib/gems/3.0.0/gems/sqlite3-1.4.2 for inspection.      
 85 Results logged to /var/lib/gems/3.0.0/extensions/aarch64-linux/3.0.0/sqlite3-1.4.2/gem_make.out
```

If you want me to retry,
& give you more output of the try,
( if more exists somewhere ),
pls tell me what to try for you.

Salut, Namaste, & Kaizen!

( :

**Comment by flavorjones** [comment]
Hello! Please watch #320 for some work related to restoring precompiled native gem support and overall better platform support. Once that's landed, if termux still isn't working, we should be in a better place to troubleshoot.

**Comment by flavorjones** [comment]
Can you please try the v1.5.0.rc1 prerelease?

**Comment by libweirdness** [comment]
Hello.

Due to the *absolute brokenness* of Ubuntu Server 22.04 LTS for both Ruby & Haskell, I have ditched Ubuntu family, entirely, & never want to see them again.

Now I'm trying Void Linux, & remembering that as much work as Slackware was, in the late 1990's, it worked.

The others I've relied on, through the years, drifted to other goals, got bought, or whatever, but Ubuntu, with their new absolute disregard for not-breaking-userland, I am removing from all systems I manage, mine & others'.

They redid the KDE spin of it, to cause a mandatory-automatic-update to make the desktop *unusable* for people who aren't geeks ( had a friend on that ), too.

People who are junkies for features, who can't care about functional integrity, are people to get the hell away from, in infrastructure.

I recommend removing Ubuntu from all systems, now.
